src/platform/i386/entry.S

```diff
@@ -110,14 +94,10 @@ restore_from_thd:
 	pushl %esp;		\
 	call fn;		\
 	addl $4, %esp;		\
-	RESTORE_REGS_GENERAL;	\
-	popl %ds; 		\
-	popl %es;		\
-	popl %fs;		\
-	popl %gs;		\
-	addl $4, %esp; 		\
-	sti;			\
-	iret
+	testl %eax, %eax;	\
+	jne restore_from_thd;	\
+	RESTORE_REGS_ALL;	\
```

Comment:
It won't be possible. Because, we are making a decision based on the return value of 'fn', Whether to restore a pre-empted thread or to return to a thread that was co-operatively switched-out of.